relay-sampling/src/lib.rs

```diff
@@ -355,6 +365,125 @@ impl TimeRange {
+    external_params: DecayingFunctionExternalParams,
+}
+
+impl DecayingFunctionExecutor {
+    fn execute(&self) -> f64 {
+        match self.decaying_fn {
+            DecayingFunction::Linear { from_sample_rate } => {
+                self.execute_linear_decay(from_sample_rate)
+            }
+            DecayingFunction::Constant => self.external_params.base_sample_rate,
+        }
+    }
+
+    /// Computes the linear decay function at a specific point in time.
+    ///
+    /// The linear decay works in the following way:
+    /// start = 100
+    /// end = 200
+    /// from = 0.7 (70% sample rate)
+    /// base_sample_rate = 0.5 (50% sample rate)
+    /// (from - base_sample_rate) = 0.2 (20% sample rate)
+    ///
+    /// If we have now = 170 it means we are at the 70% progress of the timerange, which means we
+    /// want to inversely reduce the sample rate in the interval (from - base_sample_rate).
+    /// Thus, in the 70% case, we would take only 30% of the (from - base_sample_rate) which
+    /// is equal to 0.06.
+    fn execute_linear_decay(&self, from_sample_rate: f64) -> f64 {
+        let now_timestamp = self.external_params.now.timestamp() as f64;
+        let start_timestamp = self.external_params.start.timestamp() as f64;
+        let end_timestamp = self.external_params.end.timestamp() as f64;
+
+        // We round to 2 digits in order to avoid high-precision sample rates that are more difficult
+        // to work with. Rounding is performed following the nearest integer.
+        let sample_rate_difference =
+            f64::round((from_sample_rate - self.external_params.base_sample_rate) * 100.0) / 100.0;
+        let progress = (now_timestamp - start_timestamp) / (end_timestamp - start_timestamp);
+        let inverse_progress = (1.0 - progress).clamp(0.0, 1.0);
+        self.external_params.base_sample_rate + (sample_rate_difference * inverse_progress)
+    }
+}
+
+/// A struct containing the set of external params required by a decaying function.
+#[derive(Clone, Copy, Debug)]
+struct DecayingFunctionExternalParams {
+    base_sample_rate: f64,
+    start: DateTime<Utc>,
+    end: DateTime<Utc>,
+    now: DateTime<Utc>,
+}
+
+/// A decaying function definition.
+///
+/// A decaying function is responsible of decaying the sample rate from a value to another following
+/// a given curve.
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "type")]
+pub enum DecayingFunction {
+    #[serde(rename_all = "camelCase")]
+    Linear {
+        from_sample_rate: f64,
+    },
+    Constant,
+}
+
+impl DecayingFunction {
+    /// Validate the external parameters needed by the decaying function(s).
+    fn validate_external_params(
+        &self,
+        base_sample_rate: f64,
+        time_range: TimeRange,
+        now: DateTime<Utc>,
+    ) -> Option<DecayingFunctionExecutor> {
+        if let Some((start, end)) = time_range.extract_interval_tuple() {
```

Comment:
This way, there's no need for the `extract_interval_tuple` helper function:

```suggestion
        if let (Some(start), Some(end)) = time_range {
```